File: services/engagement/src/engagement/reflection.py
"""Reflection + commitment routes — Phase 6 S57 / UX-27."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Literal
from uuid import uuid4

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy import text

from engagement.analytics.db import sessionmaker

reflection_router = APIRouter()

SCHEMA = "analytics_schema"


class ReflectionIn(BaseModel):
    user_id: str
    trigger: Literal["session", "mock", "weekly"]
    trigger_artifact_id: str | None = None
    prompt_id: str = Field(default="default_prompt", max_length=80)
    response: str | None = Field(default=None, max_length=2000)
    commitment: str | None = Field(default=None, max_length=400)
    commitment_due_at: str | None = None       # ISO timestamp


@reflection_router.post("/reflections", status_code=201)
async def post_reflection(body: ReflectionIn) -> dict:
    rid = str(uuid4())
    async with sessionmaker()() as s:
        cstatus = "pending" if body.commitment else None
        await s.execute(
            text(
                f"""
                INSERT INTO {SCHEMA}.reflections_commitments
                  (id, user_id, trigger, trigger_artifact_id, prompt_id,
                   response, commitment, commitment_due_at, commitment_status)
                VALUES
                  (CAST(:id AS uuid), CAST(:uid AS uuid), :tr,
                   CAST(:taid AS uuid), :pid, :resp, :commit,
                   CAST(:due AS timestamptz),
                   :cstatus)
                """
            ),
            {
                "id": rid,
                "uid": body.user_id,
                "tr": body.trigger,
                "taid": body.trigger_artifact_id,
                "pid": body.prompt_id,
                "resp": body.response,
                "commit": body.commitment,
                "due": body.commitment_due_at,
                "cstatus": cstatus,
            },
        )
        await s.commit()
    return {"id": rid, "trigger": body.trigger, "commitment_status": "pending" if body.commitment else None}


class CheckInIn(BaseModel):
    kept: bool
    note: str | None = Field(default=None, max_length=400)


@reflection_router.post("/commitments/{rid}/check-in")
async def check_in(rid: str, body: CheckInIn) -> dict:
    new_status = "kept" if body.kept else "missed"
    async with sessionmaker()() as s:
        res = await s.execute(
            text(
                f"""
                UPDATE {SCHEMA}.reflections_commitments
                   SET commitment_status = :st,
                       check_in_response = :note,
                       last_check_in_at = now()
                 WHERE id = CAST(:id AS uuid) AND commitment_status = 'pending'
             RETURNING id, commitment_status
                """
            ),
            {"id": rid, "st": new_status, "note": body.note},
        )
        row = res.mappings().first()
        if row is None:
            raise HTTPException(404, detail={"code": "not_found_or_already_decided"})
        await s.commit()
    return {"id": str(row["id"]), "commitment_status": row["commitment_status"]}


@reflection_router.get("/commitments/{user_id}")
async def list_commitments(user_id: str, status: str | None = None):
    where = ["user_id = CAST(:uid AS uuid)", "commitment IS NOT NULL"]
    params = {"uid": user_id}
    if status:
        where.append("commitment_status = :st")
        params["st"] = status
    async with sessionmaker()() as s:
        res = await s.execute(
            text(
                f"""
                SELECT id, trigger, prompt_id, commitment,
                       commitment_due_at, commitment_status,
                       occurred_at, last_check_in_at
                  FROM {SCHEMA}.reflections_commitments
                 WHERE {' AND '.join(where)}
              ORDER BY occurred_at DESC
                """
            ),
            params,
        )
        items = []
        for row in res.mappings():
            items.append({
                "id": str(row["id"]),
                "trigger": row["trigger"],
                "prompt_id": row["prompt_id"],
                "commitment": row["commitment"],
                "due_at": row["commitment_due_at"].isoformat() if row["commitment_due_at"] else None,
                "status": row["commitment_status"],
                "occurred_at": row["occurred_at"].isoformat(),
                "last_check_in_at": row["last_check_in_at"].isoformat() if row["last_check_in_at"] else None,
            })
    return {"user_id": user_id, "items": items}

File: apps/web-student/src/lib/reflection.ts
// Reflection + commitment client (Phase 6 S57, UX-27).
//
// Backs the reflection-and-commitment loop. Endpoints from 92efa83:
//   POST /reflections                  — record a reflection + commitment
//   POST /commitments/{rid}/check-in   — kept/missed + note
//   GET  /commitments/{user_id}?status — list commitments by status
//
// Spec: docs/02_planning/55_Phase6_UXCoPilot_Evaluation_and_SprintPlan.md S57

import { auth } from "./api";

export type ReflectionTrigger = "session" | "mock" | "weekly";
export type CommitmentStatus = "pending" | "kept" | "missed";

export interface ReflectionPayload {
  userId: string;
  trigger: ReflectionTrigger;
  triggerArtifactId?: string;
  promptId?: string;
  response?: string;
  commitment?: string;
  /** ISO timestamp the commitment is due. Optional. */
  commitmentDueAt?: string;
}

export async function postReflection(
  payload: ReflectionPayload,
): Promise<{ id: string }> {
  const body: Record<string, unknown> = {
    user_id: payload.userId,
    trigger: payload.trigger,
    prompt_id: payload.promptId ?? "default_prompt",
    response: payload.response ?? null,
    commitment: payload.commitment ?? null,
  };
  if (payload.triggerArtifactId)
    body.trigger_artifact_id = payload.triggerArtifactId;
  if (payload.commitmentDueAt) body.commitment_due_at = payload.commitmentDueAt;
  const r = await auth.fetch("/api/v1/reflections", {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!r.ok) throw new Error(`reflection post failed: HTTP ${r.status}`);
  const raw = (await r.json()) as { id: string };
  return { id: raw.id };
}

export interface Commitment {
  id: string;
  trigger: ReflectionTrigger;
  promptId: string;
  commitment: string;
  commitmentDueAt: string | null;
  commitmentStatus: CommitmentStatus;
  occurredAt: string;
  lastCheckInAt: string | null;
}

interface RawCommitment {
  id: string;
  trigger: ReflectionTrigger;
  prompt_id: string;
  commitment: string;
  commitment_due_at: string | null;
  commitment_status: CommitmentStatus;
  occurred_at: string;
  last_check_in_at: string | null;
}

export async function listCommitments(
  userId: string,
  status?: CommitmentStatus,
): Promise<Commitment[]> {
  const qs = status ? `?status=${status}` : "";
  const r = await auth.fetch(`/api/v1/commitments/${userId}${qs}`);
  if (!r.ok) throw new Error(`commitments fetch failed: HTTP ${r.status}`);
  const raw = (await r.json()) as RawCommitment[] | { items: RawCommitment[] };
  const arr = Array.isArray(raw) ? raw : (raw.items ?? []);
  return arr.map((c) => ({
    id: c.id,
    trigger: c.trigger,
    promptId: c.prompt_id,
    commitment: c.commitment,
    commitmentDueAt: c.commitment_due_at,
    commitmentStatus: c.commitment_status,
    occurredAt: c.occurred_at,
    lastCheckInAt: c.last_check_in_at,
  }));
}

export async function checkInCommitment(
  rid: string,
  kept: boolean,
  note?: string,
): Promise<CommitmentStatus> {
  const r = await auth.fetch(`/api/v1/commitments/${rid}/check-in`, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ kept, note: note ?? null }),
  });
  if (!r.ok) throw new Error(`check-in failed: HTTP ${r.status}`);
  const raw = (await r.json()) as { commitment_status: CommitmentStatus };
  return raw.commitment_status;
}
